README.md
<h3>
Mountain Menu designed by <a href='https://itcians.com'>The ITcians</a>
</h3>

<img src="video/mountain_menu.gif" width='320' height='500'>
<p>A curved menu shaped like mountain crafted for your beautiful app.
You can set your default animation and theme for this menu.

Here is an example to start with:
</p>

```dart
class _MenuDemoState extends State<MenuDemo> {
  List<String> menus = ["SETTINGS", "SEARCH", "NOTIFICATIONS","HOME","PROFILE"];
  int selectedIndex = 0;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          children: [
            Image.network(
                "https://avatars.githubusercontent.com/u/93488670?s=400&u=af911b0e99e78a599f08f47f1e3152700b869f6b&v=4"),
            const Text(
              "The ITcians",
              style: TextStyle(
                  color: Colors.black,
                  fontFamily: "Cursive",
                  fontWeight: FontWeight.bold,
                  fontSize: 40),
            ),
            Text(
              menus[selectedIndex],
              style: const TextStyle(
                  color: Color(0xff9264EB),
                  fontWeight: FontWeight.bold,
                  fontSize: 30),
            ),
          ],
        ),
      ),
      bottomNavigationBar: MountainMenu(
        color: const Color(0xff9264EB),
        menuIcon: Icons.menu,
        mainButtonElevation: 0,
        mini: false,
        curve: Curves.bounceOut,
        buttonBackgroundColor: Colors.white,
        onChange: (val) async {
          setState(() {
            selectedIndex = val;
          });
        },
        menus: const [
          Icon(
            Icons.settings,
            color: Color(0xff9264EB),
          ),
          Icon(
            Icons.search,
            color: Color(0xff9264EB),
          ),
          Icon(
            Icons.chat_bubble,
            color: Color(0xff9264EB),
          ),Icon(
            Icons.home,
            color: Color(0xff9264EB),
          ),Icon(
            Icons.person,
            color: Color(0xff9264EB),
          ),
        ],
      ),
    );
  }
}
```
<b>Properties:</b>
<ul>
<li>color: To change mountain's color</li>
<li>menuIcon: To change mountain's main icon</li>
<li>menus: List of three menu widgets</li>
<li>mini: Change the size of menu to mini and expanded</li>
<li>mainButtonElevation: To add elevation to main button</li>
<li>buttonBackgroundColor: To change background color of menu buttons</li>
<li>iconColor: To change color of menu icons</li>
<li>onChange: A menu change listener</li>
<li>curve: Animation Curve</li>
<li>borderWith: Border with of main menu button</li>
<li>duration: Animation duration</li>
<li>borderColor: To change border color of main menu</li>
</ul>


CHANGELOG.md
## 0.0.3
Menus length updated to 5 max.
You can adjust Menus size to mini or expanded. Mini property is required

test/mountain_menu_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mountain_menu/mountain_menu.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/arc.dart
library mountain_menu;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class ArcContainer extends CustomPainter {
  final Color color;

  ArcContainer({required this.color});

  @override
  void paint(Canvas canvas, Size size) {
    // var paint = Paint();
    // paint.color = Colors.red;
    // paint.style = PaintingStyle.fill;
    //
    // var path = Path();
    //
    // path.moveTo(0, size.height * 0.97);
    // path.quadraticBezierTo(size.width * 0.58, size.height * 0.675,
    //     size.width * 0.55, size.height * 0.8167);
    // path.quadraticBezierTo(size.width * 0.49, size.height * 0.675,
    //     size.width * 1.0, size.height * 1.07);
    // path.lineTo(size.width, size.height);
    // path.lineTo(0, size.height);
    //
    // canvas.drawPath(path, paint);
    Paint paint = Paint()
      ..color = color
      ..style = PaintingStyle.fill
      ..strokeWidth = 4.0;

    Path path = Path();
    path.moveTo(size.width * .35, size.height);
    path.quadraticBezierTo(
        size.width * .5, 0, size.width * .65, size.height);
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return false;
  }
}


lib/mountain_menu.dart
library mountain_menu;

import 'package:flutter/material.dart';
import 'package:mountain_menu/arc.dart';
import 'package:mountain_menu/arc_border.dart';
import 'package:mountain_menu/positions.dart';

class MountainMenu extends StatefulWidget {
  final Color color;
  final double? mainButtonElevation;
  final Color? buttonBackgroundColor;
  final Color? iconColor;
  final IconData menuIcon;
  final List<Widget> menus;
  final ValueChanged<int> onChange;
  final Curve? curve;
  final double? borderWidth;
  final Duration? duration;
  final Color? borderColor;
  final bool mini;

  const MountainMenu(
      {Key? key,
      required this.color,
      required this.menuIcon,
      required this.menus,
      required this.onChange,
      required this.mini,
      this.mainButtonElevation,
      this.borderWidth,
      this.duration,
      this.borderColor,
      this.curve,
      this.buttonBackgroundColor,
      this.iconColor})
      : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _MountainMenuState();
  }
}

class _MountainMenuState extends State<MountainMenu>
    with SingleTickerProviderStateMixin {
  bool _showMenu = false;
  late AnimationController _controller;
  final Positions _positions = Positions(); //Positions for menus

  @override
  void dispose() {
    super.dispose();
    _controller.dispose();
  }

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: widget.duration ?? const Duration(milliseconds: 700),
      vsync: this,
    );
  }

  @override
  Widget build(BuildContext context) {
    if (widget.menus.isEmpty || widget.menus.length > 5) {
      throw UnimplementedError("Menus length is not sufficient!");
    } else {
      return AnimatedContainer(
        duration: widget.duration ?? const Duration(milliseconds: 400),
        child: CustomPaint(
          painter: ArcBorder(
              color: widget.borderColor ?? Colors.yellow,
              borderWidth: widget.borderWidth),
          child: CustomPaint(
            painter: ArcContainer(color: widget.color),
            child: SizedBox(
              height: MediaQuery.of(context).size.height * .15,
              width: MediaQuery.of(context).size.width,
              child: Stack(
                children: [
                  if (_showMenu && widget.menus.length > 1)
                    Positioned(
                      top: 0,
                      left: 0,
                      right: 0,
                      child: SlideTransition(
                        position:
                            _positions.topButton().animate(CurvedAnimation(
                                  parent: _controller,
                                  curve: widget.curve != null
                                      ? widget.curve as Curve
                                      : Curves.easeInOut,
                                )),
                        child: FloatingActionButton(
                          heroTag: "menuTop",
                          backgroundColor: widget.buttonBackgroundColor,
                          mini: widget.mini,
                          onPressed: () async {
                            widget.onChange(1);
                            _animate();
                          },
                          child: widget.menus[1],
                        ),
                      ),
                    ),
                  Positioned(
                    bottom: 8,
                    left: 0,
                    right: 0,
                    child: FloatingActionButton(
                      heroTag: "mainButton",
                      elevation: widget.mainButtonElevation,
                      backgroundColor: Colors.transparent,
                      mini: widget.mini,
                      onPressed: () async {
                        _animate();
                      },
                      child: Icon(
                        _showMenu ? Icons.close : widget.menuIcon,
                        color: widget.iconColor,
                      ),
                    ),
                  ),
                  if (_showMenu && widget.menus.isNotEmpty)
                    Positioned(
                      top: MediaQuery.of(context).size.height * .06,
                      left: MediaQuery.of(context).size.width * .2,
                      child: SlideTransition(
                        position: _positions.halfLeft().animate(CurvedAnimation(
                              parent: _controller,
                              curve: widget.curve != null
                                  ? widget.curve as Curve
                                  : Curves.easeInOut,
                            )),
                        child: FloatingActionButton(
                          mini: widget.mini,
                          heroTag: "menuLeft",
                          backgroundColor: widget.buttonBackgroundColor,
                          onPressed: () async {
                            widget.onChange(0);
                            _animate();
                          },
                          child: widget.menus[0],
                        ),
                      ),
                    ),
                  if (_showMenu && widget.menus.length > 2)
                    Positioned(
                      top: MediaQuery.of(context).size.height * .06,
                      right: MediaQuery.of(context).size.width * .2,
                      child: SlideTransition(
                        position:
                            _positions.halfRight().animate(CurvedAnimation(
                                  parent: _controller,
                                  curve: widget.curve != null
                                      ? widget.curve as Curve
                                      : Curves.easeOut,
                                )),
                        child: FloatingActionButton(
                          heroTag: "menuRight",
                          mini: widget.mini,
                          backgroundColor: widget.buttonBackgroundColor,
                          onPressed: () async {
                            widget.onChange(2);
                            _animate();
                          },
                          child: widget.menus[2],
                        ),
                      ),
                    ),
                  if (_showMenu && widget.menus.length > 3)
                    Positioned(
                      left: 16,
                      bottom: 5,
                      child: SlideTransition(
                        position: _positions.left().animate(CurvedAnimation(
                              parent: _controller,
                              curve: widget.curve != null
                                  ? widget.curve as Curve
                                  : Curves.easeOut,
                            )),
                        child: FloatingActionButton(
                          heroTag: "left",
                          mini: widget.mini,
                          backgroundColor: widget.buttonBackgroundColor,
                          onPressed: () async {
                            widget.onChange(3);
                            _animate();
                          },
                          child: widget.menus[3],
                        ),
                      ),
                    ),
                  if (_showMenu && widget.menus.length > 4)
                    Positioned(
                      right: 16,
                      bottom: 5,
                      child: SlideTransition(
                        position: _positions.right().animate(CurvedAnimation(
                              parent: _controller,
                              curve: widget.curve != null
                                  ? widget.curve as Curve
                                  : Curves.easeOut,
                            )),
                        child: FloatingActionButton(
                          heroTag: "right",
                          mini: widget.mini,
                          backgroundColor: widget.buttonBackgroundColor,
                          onPressed: () async {
                            widget.onChange(4);
                            _animate();
                          },
                          child: widget.menus[4],
                        ),
                      ),
                    ),
                ],
              ),
            ),
          ),
        ),
      );
    }
  }

  void _animate() async {
    if (_showMenu) {
      await _controller.reverse();
      setState(() => _showMenu = !_showMenu);
    } else {
      setState(() => _showMenu = !_showMenu);
      await _controller.forward(from: 0);
    }
  }
}


example/lib/main.dart

import 'package:flutter/material.dart';
import 'package:mountain_menu/mountain_menu.dart';

class MenuDemo extends StatefulWidget {
  const MenuDemo({Key? key}) : super(key: key);

  @override
  State<MenuDemo> createState() => _MenuDemoState();
}

class _MenuDemoState extends State<MenuDemo> {
  List<String> menus = ["SETTINGS", "SEARCH", "NOTIFICATIONS","HOME","PROFILE"];
  int selectedIndex = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          children: [
            Image.network(
                "https://avatars.githubusercontent.com/u/93488670?s=400&u=af911b0e99e78a599f08f47f1e3152700b869f6b&v=4"),
            const Text(
              "The ITcians",
              style: TextStyle(
                  color: Colors.black,
                  fontFamily: "Cursive",
                  fontWeight: FontWeight.bold,
                  fontSize: 40),
            ),
            Text(
              menus[selectedIndex],
              style: const TextStyle(
                  color: Color(0xff9264EB),
                  fontWeight: FontWeight.bold,
                  fontSize: 30),
            ),
          ],
        ),
      ),
      bottomNavigationBar: MountainMenu(
        color: const Color(0xff9264EB),
        menuIcon: Icons.menu,
        mainButtonElevation: 0,
        mini: false,
        curve: Curves.bounceOut,
        buttonBackgroundColor: Colors.white,
        onChange: (val) async {
          setState(() {
            selectedIndex = val;
          });
        },
        menus: const [
          Icon(
            Icons.settings,
            color: Color(0xff9264EB),
          ),
          Icon(
            Icons.search,
            color: Color(0xff9264EB),
          ),
          Icon(
            Icons.chat_bubble,
            color: Color(0xff9264EB),
          ),Icon(
            Icons.home,
            color: Color(0xff9264EB),
          ),Icon(
            Icons.person,
            color: Color(0xff9264EB),
          ),
        ],
      ),
    );
  }
}
